syntax = "proto3";

package capability_registry.v1;

import "common.proto";

// Capability Registry - Tracks service capabilities and availability
// Subjects:
//   svc.capability.v1.register - Register service capabilities
//   svc.capability.v1.query - Query available capabilities
//   svc.capability.v1.heartbeat - Service heartbeat
//   evt.capability.registered.v1 - Capability registered event
//   evt.capability.removed.v1 - Capability removed event

message RegisterRequest {
  common.v1.Meta meta = 1;
  string service_id = 2;
  repeated Capability capabilities = 3;
  string endpoint = 4;  // NATS subject or URL
  map<string, string> metadata = 5;
}

message Capability {
  string capability_id = 1;
  string capability_type = 2;  // inference, generation, analysis, etc.
  repeated string supported_models = 3;
  map<string, string> parameters = 4;  // Capability-specific parameters
  double cost_multiplier = 5;  // Relative cost
  double avg_latency_ms = 6;
}

message RegisterResponse {
  common.v1.Meta meta = 1;
  bool success = 2;
  string registration_id = 3;
  int64 expires_at_ms = 4;
  common.v1.Error error = 5;
}

message QueryRequest {
  common.v1.Meta meta = 1;
  string capability_type = 2;  // Filter by type
  map<string, string> requirements = 3;  // Required parameters
  bool include_offline = 4;  // Include offline services
}

message ServiceCapability {
  string service_id = 1;
  string registration_id = 2;
  repeated Capability capabilities = 3;
  bool is_online = 4;
  int64 last_heartbeat_ms = 5;
  map<string, string> metadata = 6;
}

message QueryResponse {
  common.v1.Meta meta = 1;
  repeated ServiceCapability services = 2;
  common.v1.Error error = 3;
}

message HeartbeatRequest {
  common.v1.Meta meta = 1;
  string registration_id = 2;
  map<string, double> metrics = 3;  // Current metrics (load, latency, etc.)
}

message HeartbeatResponse {
  common.v1.Meta meta = 1;
  bool success = 2;
  int64 next_heartbeat_ms = 3;  // When next heartbeat expected
  common.v1.Error error = 4;
}

